feat(coderd): add company logo when available for email notifications

defelmnq opened the pull request:
This PR aims to close #14253 

We keep the default behavior using the Coder logo if there's no logo set.
Otherwise we want to use the logo based on the URL set in appearance.

defelmnq committed:
feat(notifications): add company logo url when available for email notifications

defelmnq committed:
feat(notifications): adapt tests with new labels

defelmnq committed:
fix: revert changes on generated code version

mafredri commented on coderd/notifications/notifier.go:
```diff
@@ -22,6 +24,10 @@ import (
 )
 
+const (
+	NotificationsDefaultLogoURL = "https://coder.com/coder-logo-horizontal.png"
```
Out of scope for this PR, but wouldn't it be better to link to a resource on the deployment here? (`https://my-deployment/logo.png`)

We can't guarantee this URL will keep working, on the deployment that would at least mean the deployment is down anyway or unreachable.

dannykopping replied:
Good idea 👍

mafredri commented on coderd/notifications/notifier.go:
```diff
@@ -223,6 +229,18 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
 
+	logoURL, err := n.store.GetLogoURL(ctx)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		n.log.Error(ctx, "failed fetching logo url", slog.Error(err))
```
Risk: We may end up showing the Coder logo here for a customer that requires it's never shown. It's better to not show a logo at all if this fails IMO.

dannykopping replied:
👍

mafredri commented on coderd/notifications/notifier.go:
```diff
@@ -223,6 +229,18 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
+	}
+
+	if logoURL == "" {
+		//nolint:ineffassign // define to default value if unable to fetch one from db
```
Is this truly an `ineffassign`, I would assume the assign to labels below takes care of it?

mafredri reviewed (commented):
Nice work! Left some feedback, but I don't have much context on decisions around this, so feel free to ignore if my feedback isn't appropriate.

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -223,6 +229,18 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
+		logoURL = NotificationsDefaultLogoURL
+	}
+
+	payload.Labels["_logo_url"] = logoURL
```
I agree with @mtojek here.

I don't think this should go into the payload. It's going to be the same for every notification, and if it's in the payload then we'll be storing the same URL for every notification.

Have a look at `coderd/notifications/dispatch/smtp/html.gotmpl: you'll see `{{ base_url }}`.
This is provided by a helper function, defined in `cli/server.go`:

```go
// templateHelpers builds a set of functions which can be called in templates.
// We build them here to avoid an import cycle by using coderd.Options in notifications.Manager.
// We can later use this to inject whitelabel fields when app name / logo URL are overridden. <----
func templateHelpers(options *coderd.Options) map[string]any {
	return map[string]any{
		"base_url":     func() string { return options.AccessURL.String() },
		"current_year": func() string { return strconv.Itoa(time.Now().Year()) },
	}
}
```

We should create a new helper which retrieves the logo instead, I think.

dannykopping commented:
> You're totally right - I just have a question as we recommend using a 3:1 aspect ratio image (`We recommend a transparent image with 3:1 aspect ratio.`)

We set `height: 40px` now, which means an image with a 3:1 aspect ratio would be 120px wide.
The [logo](https://coder.com/docs/v2/latest/api) we use is 364:81 (~4.5:1) ratio, so with the height of 40px it becomes ~180px wide.

Current:

![image](https://github.com/user-attachments/assets/8fe704b4-4013-4e78-b839-c35cb0ec8310)

Simulated with image with 3:1 aspect ratio:

![image](https://github.com/user-attachments/assets/3edd4af9-15b9-41a9-9e91-c0757dfa9957)

I think a logo with 3:1 would look quite bad (or at least, it wouldn't look like a header image; it looks quite wimpy now).

Perhaps we should use a logo of ours which is actually 3:1 and set the height a bit higher (60px) so we maintain this 180px width?

defelmnq committed:
feat(notification): move logo_url and app_name logic to helpers functions

dannykopping reviewed (commented):
Nearly there, good work!

defelmnq committed:
chore: remove unused GetLogoURL method from notifications store interface

mafredri commented on cli/server.go:
```diff
@@ -1309,6 +1309,30 @@ func templateHelpers(options *coderd.Options) map[string]any {
 		"base_url":     func() string { return options.AccessURL.String() },
 		"current_year": func() string { return strconv.Itoa(time.Now().Year()) },
+		"logo_url": func() string {
+			logoURL, err := options.Database.GetLogoURL(context.Background())
```
Perhaps passing a context to `templateHelpers` (server lifetime) would be a good idea, and perhaps we should define a maximum execution time here (context.WithTimeout) so that we don't end up in a situation where the DB query simply blocks for an indefinite amount of time.

mtojek replied:
👍 

I have similar feelings. We might be risking by placing DB call inside a template helper. I'm wondering if we shouldn't implement something similar to AppearanceFetcher for the site handler.

dannykopping replied:
Good catch 👍

mtojek replied:
What would you say if we slightly modify this concept. My main concern is around swallowing the error in favor of returning an empty string `""`, so:

1. Wherever we render the `htmlBody`:

```go
	htmlBody, err = render.GoTemplate(htmlTemplate, payload, s.helpers)
	if err != nil {
		return nil, xerrors.Errorf("render full html template: %w", err)
	}
```

we create a dynamic `template.FuncMap` with `app_name` and `logo_url`. We could add a reference to the store to `SMTPHandler`, and retrieve values before rendering.

Thoughts?

mafredri reviewed (approved):
Except for the background contexts and potential for locking up through stalled DB queries, I think this looks great, nice work!

dannykopping commented on coderd/notifications/dispatch/smtp/html.gotmpl:
```diff
@@ -8,7 +8,10 @@
       <div style="text-align: center;">
-        <img src="https://coder.com/coder-logo-horizontal.png" alt="Coder Logo" style="height: 40px;" />
+        <img src="{{ logo_url }}" alt="Company Logo" style="height: 40px;" />
+        <h1 style="text-align: center; font-size: 36px; font-weight: 400; margin: 8px 0 32px; line-height: 1.5;">
```
I don't think the name is necessary except for use as alt-text when the logo is missing.

defelmnq committed:
fix: add better context and timeout for db related queries

defelmnq committed:
chore: lint

mtojek commented on coderd/notifications/notifier.go:
```diff
@@ -223,6 +229,18 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
 
+	logoURL, err := n.store.GetLogoURL(ctx)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		n.log.Error(ctx, "failed fetching logo url", slog.Error(err))
```
This could be an option to make it less awkward: https://github.com/coder/coder/pull/14935#discussion_r1787508173

... and just retry.

dannykopping replied:
Errors in retrieving the data from the db will cause the notification to fail, and it'll be automatically retried later.
If we have a persistent issue which causes the notifications to exceed their retry limit (5 by default) then that's actually a good signal that something is wrong.

defelmnq committed:
feat(notifications): improve logo_url and app_name fetching moving it to notifications package

defelmnq committed:
feat(notifications): improve logo_url and app_name fetching moving it to notifications package

defelmnq committed:
feat(notifications): improve logo_url and app_name fetching moving it to notifications package

defelmnq committed:
feat(notifications): improve logo_url and app_name fetching moving it to notifications package

Commit:
feat(coderd): regenerate golden files

defelmnq commented:
Hey @mtojek - based on our conversation at the beginning of the week I changed the logic to fetch things inside the notifications package and remove it from the cli one. I think this logic should match more with what you had in mind, and also helps us to use both the context and errors ! 🔥

mtojek commented on coderd/notifications/testdata/rendered-templates/smtp/TemplateTemplateDeleted.html.golden:
```diff
@@ -37,8 +37,12 @@ l', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif; color: #020617=
       <div style=3D"text-align: center;">
-        <img src=3D"https://coder.com/coder-logo-horizontal.png" alt=3D"Cod=
-er Logo" style=3D"height: 40px;" />
+        <img src=3D"https://coder.com/coder-logo-horizontal.png" alt=3D"Log=
```
Can we set alternative text to the app name?

defelmnq replied:
I changed :

- Added the app_name as the alt of img.
- Remove the app_name from the email header as discussed with @dannykopping

mtojek commented on coderd/notifications/testdata/rendered-templates/smtp/TemplateUserAccountCreated.html.golden:
```diff
@@ -39,8 +39,12 @@ l', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif; color: #020617=
       <div style=3D"text-align: center;">
-        <img src=3D"https://coder.com/coder-logo-horizontal.png" alt=3D"Cod=
-er Logo" style=3D"height: 40px;" />
+        <img src=3D"https://coder.com/coder-logo-horizontal.png" alt=3D"Log=
```
Looks like we're missing tests verifying this change. Can you check if it is easy to pick one SMTP template and render it using a custom logo/app name? It would be cool to have at least one golden file.

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -223,6 +230,30 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
 		return nil, xerrors.Errorf("failed to resolve handler %q", msg.Method)
 	}
 
+	appName, err := n.store.GetApplicationName(ctx)
```
I think this should get moved into its own function.
You'll then be able to add an internal test to validate that the helpers return the correct thing when the db changes.

defelmnq committed:
feat(notifications): working on tests for logo and app_name

Commit:
feat(notifications): change helpers logic to be defined in the Dispatcher function

Commit:
feat(notifications): add golden file for custom appearance

Commit:
feat(notifications): add golden file for custom appearance

defelmnq commented:
Update here - based on the last conversation with the team I moved the `fetchApplicationName` and `fetchLogoURL` functions to be isolated.

Also generated a specific golden file for a notification with custom application name and logo URL so we can validate they are correctly fetched.

After discussing it with @dannykopping I changed the Dispatcher signature to take dynamic helpers instead of ones defined at instantiation.

Finally, in term of logic we have three paths : 
- Either the appearance settings are not defined and we take Coder values.
- Either they are defined and we take it.
- Either we have a DB error while trying to fetch it and we go to the error path without sending the notification.

mtojek commented on coderd/notifications/metrics_test.go:
```diff
@@ -43,7 +44,7 @@ func TestMetrics(t *testing.T) {
 	reg := prometheus.NewRegistry()
 	metrics := notifications.NewMetrics(reg)
-	template := notifications.TemplateWorkspaceDeleted
+	tmplate := notifications.TemplateWorkspaceDeleted
```
typo: template?

defelmnq replied:
renamed to tmpl due to package import conflict.

mtojek commented on coderd/notifications/notifications_test.go:
```diff
@@ -1610,3 +1611,183 @@ func (n *acquireSignalingInterceptor) AcquireNotificationMessages(ctx context.Co
+	err = smtpManager.Stop(ctx)
+	require.NoError(t, err)
+
+	goldenFile := filepath.Join("testdata", "rendered-templates", "smtp", "TemplateWorkspaceDeleted_WithCustomAppearance.html.golden")
```
nit: t.Name() + `.html.golden`?

mtojek commented on coderd/notifications/notifications_test.go:
```diff
@@ -1610,3 +1611,183 @@ func (n *acquireSignalingInterceptor) AcquireNotificationMessages(ctx context.Co
+	err := db.UpsertApplicationName(ctx, "CustomApplication")
+	require.NoError(t, err)
+
+	err = db.UpsertLogoURL(ctx, "https://custom.application")
```
nit: actually, you could use a link that _at least_ pretends to be PNG :)

mtojek commented on coderd/notifications/notifications_test.go:
```diff
@@ -1610,3 +1611,183 @@ func (n *acquireSignalingInterceptor) AcquireNotificationMessages(ctx context.Co
+	// nolint:gocritic // Unit test.
+	ctx := dbauthz.AsSystemRestricted(testutil.Context(t, testutil.WaitSuperLong))
+
+	err := db.UpsertApplicationName(ctx, "CustomApplication")
```
nit: Custom Application (space bar)

PS. The more natural golden files look, the better for testing coderd logic.

mtojek commented on coderd/notifications/notifications_test.go:
```diff
@@ -1610,3 +1611,183 @@ func (n *acquireSignalingInterceptor) AcquireNotificationMessages(ctx context.Co
+		t.Skip("This test requires postgres; it relies on the notification templates added by migrations in the database")
+	}
+
+	const (
+		username = "bob"
+		password = "🤫"
+
+		hello = "localhost"
+
+		from = "system@coder.com"
+		hint = "run \"DB=ci make update-golden-files\" and commit the changes"
+	)
```
maybe extract a common one from `TestNotificationTemplates_Golden`?

mtojek commented on coderd/notifications/notifications_test.go:
```diff
@@ -1610,3 +1611,183 @@ func (n *acquireSignalingInterceptor) AcquireNotificationMessages(ctx context.Co
+			},
+		)
+		return api.Database, &api.Logger, &user
+	}()
```
Ok, let's start with this comment - 

what differentiates your test from other tests in `TestNotificationTemplates` is setting the app name & logo.

I think what you need to do and safe over hundred LOC is to extend:

```
tests := []struct {
		name    string
		id      uuid.UUID
		payload types.MessagePayload
	}
```

with 

```
tests := []struct {
		name    string
		id      uuid.UUID
		payload types.MessagePayload

                 appName string
                 logoURL string
	}
``` 

this way you could use the same testing logic, just extended with API call to set app name and logo URL.

defelmnq replied:
I merged tests as discussed - makes way more sense and clean up the other test 👌

defelmnq committed:
feat(notifications): work on tests

defelmnq committed:
feat(notifications): add golden file for custom appearance

defelmnq committed:
feat(notifications): add golden file for custom appearance

mtojek reviewed (approved):
Ship it 👍

defelmnq committed:
feat(notifications): add comment ent in tests for enterprise feature

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -230,39 +228,32 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
-			return nil, xerrors.Errorf("get application name: %w", err)
-		}
-		appName = notificationsDefaultAppName
+	helpers := make(template.FuncMap)
```
Nit: all of this helper code should be moved out to a function.

dannykopping commented on coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceDeleted_CustomAppearance.html.golden:
Nice 👍

dannykopping reviewed (commented):
Nearly there!

defelmnq committed:
feat(coderd): remove unused call

johnstcn commented on coderd/notifications/dispatch/smtp/html.gotmpl:
```diff
@@ -8,7 +8,7 @@
     <div style="max-width: 600px; margin: 20px auto; padding: 60px; border: 1px solid #e2e8f0; border-radius: 8px; background-color: #fff; text-align: left; font-size: 14px; line-height: 1.5;">
       <div style="text-align: center;">
-        <img src="https://coder.com/coder-logo-horizontal.png" alt="Coder Logo" style="height: 40px;" />
+        <img src="{{ logo_url }}" alt="{{ app_name }} Logo" style="height: 40px;" />
```
note: my immediate thought was to use a [Data URL](https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data) but apparently support for this is extremely limited.

johnstcn commented on coderd/notifications/testdata/rendered-templates/smtp/TemplateWorkspaceDeleted_CustomAppearance.html.golden:
```diff
@@ -0,0 +1,90 @@
+<p>The specified reason was &ldquo;<strong>autodeleted due to dormancy (aut=
+obuild)</strong>&rdquo;.</p>
+      </div>
+      <div style=3D"text-align: center; margin-top: 32px;">
```
> style=3D

This smells like bad URL encoding to me?

dannykopping replied:
I think it's ok:
https://en.wikipedia.org/wiki/Quoted-printable#Quoted-printable_encoding

johnstcn replied:
<obligatory link to XKCD#927>

johnstcn commented on coderd/notifications/notifier.go:
```diff
@@ -22,6 +22,11 @@ import (
 )
 
+const (
+	notificationsDefaultLogoURL = "https://coder.com/coder-logo-horizontal.png"
```
Can we instead link to something hosted on the Coder deployment itself?

dannykopping replied:
I believe @defelmnq created a follow-up for this already?

johnstcn replied:
Ah thanks, I must have missed it in the above discussion.

defelmnq committed:
fix(notifications): improve tests and some nit fixes

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -228,22 +228,10 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
+	helpers, err := n.fetchHelpers(ctx)
 	if err != nil {
-		return nil, xerrors.Errorf("fetch app name: %w", err)
+		return nil, xerrors.Errorf("fetch helpers: %w", err)
```
Have you validated that failures here cause a "temporary failure" so that subsequent requests are retried?

dannykopping reviewed (approved):
LGTM!

We just need to ensure failures to fetch app name / logo will be retried, then we can get this merged.

stirby commented:
Exciting!

defelmnq committed:
chore(retry): improve retry policy on fetcher

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -168,7 +172,11 @@ func (n *notifier) process(ctx context.Context, success chan<- dispatchResult, f
 		if err != nil {
 			n.log.Warn(ctx, "dispatcher construction failed", slog.F("msg_id", msg.ID), slog.Error(err))
-			failure <- n.newFailedDispatch(msg, err, false)
+			if xerrors.Is(err, errFetchfailed) {
+				failure <- n.newFailedDispatch(msg, err, true)
+			} else {
+				failure <- n.newFailedDispatch(msg, err, false)
+			}
```
```suggestion
		// Failure to prepare a notification is usually fatal and cannot be retried, but there is a case where fetching whitelabelling info (app name / logo) can fail due to a temporary database hiccup which should be retried.
		failure <- n.newFailedDispatch(msg, err, errors.Is(err, errDecorateHelpersFailed))
```

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -27,6 +27,10 @@ const (
 )
 
+var (
+	errFetchfailed = xerrors.New("failed to fetch helpers")
```
```suggestion
	errDecorateHelpersFailed = xerrors.New("failed to decorate helpers")
```

dannykopping commented on coderd/notifications/notifier.go:
```diff
@@ -230,7 +238,7 @@ func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotification
 	helpers, err := n.fetchHelpers(ctx)
 	if err != nil {
-		return nil, xerrors.Errorf("fetch helpers: %w", err)
+		return nil, errFetchfailed
```
`fetchHelpers` is not a very descriptive name. What you're actually doing is decorating the helpers with additional information.

```suggestion
		return nil, errDecorateHelpersFailed
```

defelmnq committed:
Merge remote-tracking branch 'origin/main' into feat/notif-custom-logo

defelmnq committed:
improve errors handling

defelmnq committed:
improve errors handling

defelmnq committed:
improve errors handling
